build: migrate to conan 2

sublimator opened the pull request:
# PR #585: Migrate to Conan 2

## Summary

This PR migrates the xahaud build system from Conan 1 to Conan 2, addressing deprecation warnings and preparing for Conan 1's end-of-life. The migration maintains full compatibility with both local development and Holy Build Box (HBB) release builds while simplifying several aspects of the build configuration.

## Key Changes

### 1. Conan 2 Syntax and Configuration

#### Updated Command Syntax
- **Export commands**: Changed from `conan export <path> <reference>` to `conan export <path> --version <ver> --user <user> --channel <channel>`
- **Profile management**: Migrated from Conan 1 profile format to Conan 2's cleaner syntax
- **Cache paths**: Updated from `/cache/conan` to `/cache/conan2` with separate download and source caches
- **Version overrides**: Ensure consistency with `override=True` for boost/1.86.0, sqlite3/3.42.0, snappy/1.1.10

#### Profile Structure (Conan 2 Format)
```ini
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux

[conf]
# Build tools configuration
tools.build:compiler_executables={"c": "/usr/bin/gcc", "cpp": "/usr/bin/g++"}
```

### 2. Dual Boost Configuration Documentation

The PR clarifies and documents the intentional dual Boost setup in HBB builds:

- **Manual Boost (minimal)**: Built in Docker for WasmEdge only (filesystem + system components)
- **Conan Boost (full)**: Provides all components needed by the application and dependencies (nudb, soci)

Both are version 1.86.0, ensuring symbol compatibility. The static linking resolves any potential conflicts at link time.

**Why not just use Conan's Boost for everything?** WasmEdge is compiled during the Docker image build (for caching efficiency), while Conan runs later inside a container using that image. The Docker-cached WasmEdge build needs Boost libraries to be present at Docker build time, before Conan has even run.

**Why not manage everything through Conan?** This is technically possible but would require:
- A custom WasmEdge recipe that builds from source (upstream only provides prebuilt binaries)
- A custom LLVM 14.x recipe (Conan Center jumped from 13.x to 19.x, skipping the version WasmEdge needs)
- Figuring out how to make Conan recipes use system libraries (not straightforward in Conan's isolated build environment)

The current approach of Docker-cached manual builds + Conan for standard dependencies is a pragmatic compromise that avoids maintaining complex custom recipes.

### 3. Build Tool Requirements

Added `tool_requires()` for build tools to ensure proper glibc compatibility:
- `protobuf/3.21.12` - Provides protoc compiler
- `grpc/1.50.1` - Provides gRPC plugins
- `b2/5.3.2` (HBB only) - Forces source build to avoid GLIBCXX version issues

Additionally, the HBB Docker configuration includes:
```ini
[conf]
# Force building from source for packages with binary compatibility issues
*:tools.system.package_manager:mode=build
```
This ensures all Conan packages are built from source with HBB's gcc-toolset-11, avoiding prebuilt binaries that might have incompatible glibc requirements.

### 4. Updated Custom WasmEdge Recipe

Updated `external/wasmedge/conanfile.py` to handle prebuilt WasmEdge binaries:
- Updated `package_id()` implementation for Conan 2

### 5. CI/CD Updates

All GitHub Actions workflows updated for Conan 2:
- **Linux**: Full Conan 2 profile configuration
- **macOS**: CMake 3.31.7 for compatibility, Apple Clang workarounds
- **Docker**: Comprehensive dependency installation with proper caching


## Technical Details

### HBB Build Environment

The Holy Build Box environment requires special handling for glibc 2.28 compatibility:
- Prebuilt b2 binaries require GLIBCXX_3.4.29 (too new for HBB's glibc 2.28)
- Solution: Force b2 to build from source via `tool_requires('b2/5.3.2')`
- Manual LLVM 14 and WasmEdge installation for static linking

**Key constraint**: WasmEdge requires LLVM 14.x specifically for WebAssembly compilation, but Conan Center only offers 11.x, 12.x, 13.x, or 19.x. This gap necessitates manual installation.

### Build Verification

Added comprehensive build verification:
- Verbose final link step to debug symbol resolution
- Full `ldd` output to verify dynamic dependencies
- `libcheck` validation for system library compliance

### Temporary File Handling

Improved temp file management in containerized builds:
```bash
export TMPDIR=/io/tmp
export TEMP=/io/tmp
export TMP=/io/tmp
```

## Migration Benefits

1. **Future-proof**: Conan 1 reaches end-of-life soon
2. **Cleaner syntax**: Conan 2's profile and configuration format is more maintainable
3. **Better caching**: Separate download/source caches improve build performance
4. **Explicit dependencies**: `tool_requires()` clearly separates build tools from library dependencies
5. **Improved documentation**: Dual Boost setup is now clearly explained

## Testing

- ✅ Local development builds (macOS, Linux)
- ✅ HBB release builds with glibc 2.28 compatibility
- ✅ CI/CD pipelines (all platforms)
- ✅ Static linking verification via `ldd` and `libcheck`

## Breaking Changes

None. The migration maintains full backwards compatibility:
- Same compiler requirements
- Same dependency versions (except minor bumps for bug fixes)
- Same build outputs
- Same binary compatibility

## Notes for Reviewers

1. The dual Boost setup in HBB is intentional and documented - both versions are 1.86.0
2. The `tool_requires('b2/5.3.2')` in HBB is necessary to avoid glibc incompatibility
3. The WasmEdge recipe handles prebuilt binaries only (no source build option from upstream)

## Checklist

- [x] All tests pass
- [x] HBB release builds work
- [x] Backwards compatibility maintained

sublimator committed:
build: migrate to Conan 2

Migrates the build system from Conan 1 to Conan 2:

- Update all conan commands to v2 syntax (export, install, profile)
- Convert profiles to Conan 2 INI format with [settings] and [conf] sections
- Add tool_requires() for build tools (protobuf, grpc, b2)
- Fix wasmedge/0.11.2 recipe for Conan 2 compatibility (package_id syntax)
- Update CI workflows for all platforms (Linux, macOS, Docker/HBB)
- Document dual Boost setup (manual for WasmEdge, Conan for app)
- Force glibc-compatible builds via source compilation in HBB
- Ensure dependency version consistency with override=True

Maintains full backwards compatibility while preparing for Conan 1 EOL.

sublimator committed:
refactor: remove redundant dual boost comment

The comment between configure() and requirements() was redundant
since the detailed explanation in requirements() is more comprehensive.

sublimator committed:
fix: add set -ex to release-builder.sh for fail-fast behavior

Enable bash strict mode to:
- Exit immediately on command failure (set -e)
- Print commands before execution for debugging (set -x)

This matches the documented behavior in the script comments.

dangell7 commented:
We will need to update the BUILD.md file as well. Conan 1 commands don't work for Conan 2.

dangell7 commented:
Can you update wasmedge to solve those other issues? 

We didn't do wasm like that for caching we did it like that because it was just easier to install it and link then to build from source in cmake. When we first were developing we installed wasm locally and linked it there was no cake build. So I don't think there's any reason not to just use Conan for everything now, if you can.

sublimator commented:
> We didn't do wasm like that for caching 

I might not have been clear there. "WasmEdge is compiled during the Docker image build (for caching efficiency". What that was trying to say is that it's done in the (cached) Docker image build step. If it were done inside the "release-builder.sh" script it would be slow. Conan installed boost is not available at that point. So that's why it's not trivial to simply use conan boost to compile wasmedge. 

Your next question might be: "So why not just compile boost AND Wasm edge using Conan?" 

There's no recipe available for wasmedge to actually build it from source. It has dependencies on llvm and the llvm linker and boost. LLVM itself, at version 14, is not available on conan. I do not know about the linker. 

> So I don't think there's any reason not to just use Conan for everything now, if you can.

This is not news though. It's what was happening in the Conan 1 build. We're only really special casing things for the HBB build, and we're using the Conan wasmedge recipe. 

We CAN use conan for everything (nothing is impossible), but we'll need to make custom recipes for wasmedge that don't just download prebuilt binaries. We'll also have to manage the dependencies that we're currently installing in HBB, including llvm and the linker.

> Key constraint: WasmEdge requires LLVM 14.x specifically for WebAssembly compilation, but Conan Center only offers 11.x, 12.x, 13.x, or 19.x. This gap necessitates manual installation.

As per ^ 

I didn't really think it was worth it.

dangell7 commented:
> > We didn't do wasm like that for caching
> 
> I might not have been clear there. "WasmEdge is compiled during the Docker image build (for caching efficiency". What that was trying to say is that it's done in the (cached) Docker image build step. If it were done inside the "release-builder.sh" script it would be slow. Conan installed boost is not available at that point. So that's why it's not trivial to simply use conan boost to compile wasmedge.
> 
> Your next question might be: "So why not just compile boost AND Wasm edge using Conan?"
> 
> There's no recipe available for wasmedge to actually build it from source. It has dependencies on llvm and the llvm linker and boost. LLVM itself, at version 14, is not available on conan. I do not know about the linker.
> 
> > So I don't think there's any reason not to just use Conan for everything now, if you can.
> 
> This is not news though. It's what was happening in the Conan 1 build. We're only really special casing things for the HBB build, and we're using the Conan wasmedge recipe.
> 
> We CAN use conan for everything (nothing is impossible), but we'll need to make custom recipes for wasmedge that don't just download prebuilt binaries. We'll also have to manage the dependencies that we're currently installing in HBB, including llvm and the linker.
> 
> > Key constraint: WasmEdge requires LLVM 14.x specifically for WebAssembly compilation, but Conan Center only offers 11.x, 12.x, 13.x, or 19.x. This gap necessitates manual installation.
> 
> As per ^
> 
> I didn't really think it was worth it.

This recipe has a llvm dependency? https://conan.io/center/recipes/wasmedge?version=0.11.2? Yeah maybe its HBB but I've built xahau using this recipe before. 

But ok doing it like this is fine.

sublimator commented:
> This recipe has a llvm dependency? https://conan.io/center/recipes/wasmedge?version=0.11.2? > Yeah maybe its HBB but I've built xahau using this recipe before.

Yeah, that works outside of HBB just fine, but all it does is download pre-built binaries (which aren't suitable for HBB, being shared libs that also link to other shared libs)

sublimator commented:
> We will need to update the BUILD.md file as well. Conan 1 commands don't work for Conan 2.

On it !

sublimator committed:
docs: update BUILD.md for conan 2 migration

- Update minimum requirements from Conan 1.55 to Conan 2.x
- Replace conan profile new/update commands with Conan 2 syntax
- Update export commands to use --version --user --channel flags
- Add WasmEdge export step
- Update profile configuration to use direct file editing at ~/.conan2/profiles/default
- Update troubleshooting section with Conan 2 cache locations and commands
- Replace outdated BOOST_ASIO_HAS_STD_INVOKE_RESULT workaround with Apple Clang 17+ gRPC fix
- Fix Conan downloads URL

tequdev commented on release-builder.sh:
```diff
@@ -90,29 +92,37 @@ RUN /hbb_exe/activate-exec bash -c "dnf install -y epel-release && \
-# Install Conan and CMake
-RUN /hbb_exe/activate-exec pip3 install "conan==1.66.0" && \
+# Install Conan 2 and CMake
+RUN /hbb_exe/activate-exec pip3 install "conan>=2.0" && \
```
In this specify, wouldn't conan3 be installed when conan3 is released??

sublimator replied:
Oooh, ooops, thanks !

tequdev commented on conanfile.py:
```diff
@@ -23,22 +23,27 @@ class Xrpl(ConanFile):
+    import os
+
+    # Check if we're in HBB release build environment
+    IS_HBB_BUILD = os.getenv('HBB_RELEASE_BUILD') == '1'
```
It may be clearer to simply use the variable name `IS_STATIC_BUILD`.

HBB is a tool for static builds.

sublimator replied:
I would argue IS_STATIC_BUILD would be too generic and misleading. It's specifically for changing the requirements when building for HBB.

It also controls workarounds to build some components from source for older glibc versions, using tool_requires.

It's simple and you know exactly what it's used for. As you know, the HBB build messes around with overriding cmake modules and doing sed surgeries. 

Simply setting HBB_RELEASE_BUILD -> IS_STATIC_BUILD would not be enough, so I think the existing name is actually more appropriate, thanks.

tequdev replied:
I didn't like that the name of a tool only used in release-builder was being used here. However, it's fine as is.

sublimator replied:
Yeah, I feel ya, the conanfile knowing about some concrete build environment is kind of strange, but it's self-documenting vs having to send a bunch of more abstract yet "pure" parameters that we'd likely only ever really use in HBB anyway. 

At the end of the day, it's a python file for a reason, and it's not THAT much of a stretch from checking other environment context like the operating system etc

sublimator replied:
I'll have a quick look at some alternatives while I'm here anyway

sublimator replied:
Had a look at the alternatives. Here are the options:

**Current approach (env var):**
```python
IS_HBB_BUILD = os.getenv('HBB_RELEASE_BUILD') == '1'

def requirements(self):
    if not IS_HBB_BUILD:
        requires.append('wasmedge/0.11.2@xahaud/stable')

def build_requirements(self):
    if IS_HBB_BUILD:
        self.tool_requires('b2/5.3.2')
```
Usage: `export HBB_RELEASE_BUILD=1`

**Alternative 1 (Conan option - unified):**
```python
options = {'hbb_mode': [True, False]}
default_options = {'hbb_mode': False}

def requirements(self):
    if not self.options.hbb_mode:
        self.requires('wasmedge/0.11.2@xahaud/stable')

def build_requirements(self):
    if self.options.hbb_mode:
        self.tool_requires('b2/5.3.2')
```
Usage: `conan install .. -o hbb_mode=True`

**Alternative 2 (separate options):**
```python
options = {
    'with_wasmedge': [True, False],
    'tool_requires_b2': [True, False]
}

def requirements(self):
    if self.options.with_wasmedge:
        self.requires('wasmedge/0.11.2@xahaud/stable')

def build_requirements(self):
    if self.options.tool_requires_b2:
        self.tool_requires('b2/5.3.2')
```
Usage: `conan install .. -o with_wasmedge=False -o tool_requires_b2=True`

What do you think?

Let me look at #2

sublimator replied:
Option 2 impl in [c470ab2](https://github.com/Xahau/xahaud/pull/585/commits/c470ab2169e107aa996443c40d7518d9fc9d037a)

sublimator committed:
fix: constrain conan version to 2.x in release-builder

Prevent potential breakage from future Conan 3.x by explicitly
constraining pip install to "conan>=2.0,<3.0"

sublimator committed:
refactor: replace HBB env var with conan options

Replace HBB_RELEASE_BUILD environment variable with explicit Conan options
to decouple conanfile.py from specific build environments.

- Add 'with_wasmedge' and 'tool_requires_b2' options to conanfile.py
- Remove HBB_RELEASE_BUILD env var check from conanfile.py
- Pass options explicitly in build-core.sh: -o with_wasmedge=False -o tool_requires_b2=True
- Remove HBB_RELEASE_BUILD=1 export from build-full.sh
- Clean up HBB-specific commentary (already documented in release-builder.sh)

RichardAH committed:
Merge branch 'dev' into nd-migrate-to-conan-2-2025-09-22
